File: packages/flutter-im-ui/test/form_behavior_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flare_im_ui/flare_im_ui.dart';

// DoD 20 —— spec/form-keyboard-contract.json 的向量表。四端跑同一批 id，
// tooling/check-form-keyboard.mjs 会在某一端不再验某条时报红。
typedef _Case = ({
  String id,
  String key,
  bool shift,
  bool composing,
  bool popupOpen,
  bool submitOnEnter,
  FlareFormKeyboardIntent intent,
});

_Case _c(
  String id,
  String key,
  FlareFormKeyboardIntent intent, {
  bool shift = false,
  bool composing = false,
  bool popupOpen = false,
  bool submitOnEnter = false,
}) => (
  id: id,
  key: key,
  shift: shift,
  composing: composing,
  popupOpen: popupOpen,
  submitOnEnter: submitOnEnter,
  intent: intent,
);

void main() {
  final vectors = <_Case>[
    _c('enter.submits', 'Enter', FlareFormKeyboardIntent.submit, submitOnEnter: true),
    _c('enter.withoutSubmitMode', 'Enter', FlareFormKeyboardIntent.none),
    _c('enter.composing', 'Enter', FlareFormKeyboardIntent.none, composing: true, submitOnEnter: true),
    _c('tab.forward', 'Tab', FlareFormKeyboardIntent.tabForward),
    _c('tab.backward', 'Tab', FlareFormKeyboardIntent.tabBackward, shift: true),
    _c('tab.composing', 'Tab', FlareFormKeyboardIntent.none, composing: true),
    _c('escape.closesPopup', 'Escape', FlareFormKeyboardIntent.closePopup, popupOpen: true),
    _c('escape.withoutPopup', 'Escape', FlareFormKeyboardIntent.none),
    _c('escape.composing', 'Escape', FlareFormKeyboardIntent.none, composing: true, popupOpen: true),
    _c('arrowDown.nextOption', 'ArrowDown', FlareFormKeyboardIntent.nextOption, popupOpen: true),
    _c('arrowUp.previousOption', 'ArrowUp', FlareFormKeyboardIntent.previousOption, popupOpen: true),
    _c('arrowDown.composing', 'ArrowDown', FlareFormKeyboardIntent.none, composing: true, popupOpen: true),
    _c('arrowDown.withoutPopup', 'ArrowDown', FlareFormKeyboardIntent.none),
    _c('home.firstOption', 'Home', FlareFormKeyboardIntent.firstOption, popupOpen: true),
    _c('end.lastOption', 'End', FlareFormKeyboardIntent.lastOption, popupOpen: true),
    _c('pageDown.pageForward', 'PageDown', FlareFormKeyboardIntent.pageForward, popupOpen: true),
    _c('pageUp.pageBackward', 'PageUp', FlareFormKeyboardIntent.pageBackward, popupOpen: true),
    _c('unknownKey', 'F13', FlareFormKeyboardIntent.none, popupOpen: true),
  ];

  for (final v in vectors) {
    test(v.id, () {
      expect(
        resolveFormKeyboardIntent(
          v.key,
          shift: v.shift,
          composing: v.composing,
          popupOpen: v.popupOpen,
          submitOnEnter: v.submitOnEnter,
        ),
        v.intent,
      );
    });
  }

  test('输入法组字时每个键都归输入法', () {
    // 守卫是无条件的：组字中的 Tab 不能把焦点从半个词上挪走，Escape 也属于输入法。
    for (final key in [
      'Enter',
      'Tab',
      'Escape',
      'ArrowDown',
      'ArrowUp',
      'Home',
      'End',
      'PageDown',
      'PageUp',
    ]) {
      expect(
        resolveFormKeyboardIntent(
          key,
          shift: true,
          composing: true,
          popupOpen: true,
          submitOnEnter: true,
        ),
        FlareFormKeyboardIntent.none,
        reason: key,
      );
    }
  });

  test('键名大小写不敏感', () {
    expect(
      resolveFormKeyboardIntent('enter', submitOnEnter: true),
      FlareFormKeyboardIntent.submit,
    );
    expect(
      resolveFormKeyboardIntent('ARROWDOWN', popupOpen: true),
      FlareFormKeyboardIntent.nextOption,
    );
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/FormBehaviorTest.kt
package com.flare.im.ui

import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * DoD 20 — the vector table of spec/form-keyboard-contract.json. The four
 * platforms run the same ids; tooling/check-form-keyboard.mjs fails when one of
 * them stops exercising a vector.
 */
class FormBehaviorTest {
    private data class Vector(
        val id: String,
        val key: String,
        val intent: FlareFormKeyboardIntent,
        val shift: Boolean = false,
        val composing: Boolean = false,
        val popupOpen: Boolean = false,
        val submitOnEnter: Boolean = false,
    )

    private val vectors = listOf(
        Vector("enter.submits", "Enter", FlareFormKeyboardIntent.Submit, submitOnEnter = true),
        Vector("enter.withoutSubmitMode", "Enter", FlareFormKeyboardIntent.None),
        Vector("enter.composing", "Enter", FlareFormKeyboardIntent.None, composing = true, submitOnEnter = true),
        Vector("tab.forward", "Tab", FlareFormKeyboardIntent.TabForward),
        Vector("tab.backward", "Tab", FlareFormKeyboardIntent.TabBackward, shift = true),
        Vector("tab.composing", "Tab", FlareFormKeyboardIntent.None, composing = true),
        Vector("escape.closesPopup", "Escape", FlareFormKeyboardIntent.ClosePopup, popupOpen = true),
        Vector("escape.withoutPopup", "Escape", FlareFormKeyboardIntent.None),
        Vector("escape.composing", "Escape", FlareFormKeyboardIntent.None, composing = true, popupOpen = true),
        Vector("arrowDown.nextOption", "ArrowDown", FlareFormKeyboardIntent.NextOption, popupOpen = true),
        Vector("arrowUp.previousOption", "ArrowUp", FlareFormKeyboardIntent.PreviousOption, popupOpen = true),
        Vector("arrowDown.composing", "ArrowDown", FlareFormKeyboardIntent.None, composing = true, popupOpen = true),
        Vector("arrowDown.withoutPopup", "ArrowDown", FlareFormKeyboardIntent.None),
        Vector("home.firstOption", "Home", FlareFormKeyboardIntent.FirstOption, popupOpen = true),
        Vector("end.lastOption", "End", FlareFormKeyboardIntent.LastOption, popupOpen = true),
        Vector("pageDown.pageForward", "PageDown", FlareFormKeyboardIntent.PageForward, popupOpen = true),
        Vector("pageUp.pageBackward", "PageUp", FlareFormKeyboardIntent.PageBackward, popupOpen = true),
        Vector("unknownKey", "F13", FlareFormKeyboardIntent.None, popupOpen = true),
    )

    @Test fun everyVectorResolvesToItsIntent() {
        for (v in vectors) {
            assertEquals(
                v.intent,
                resolveFormKeyboardIntent(v.key, shift = v.shift, composing = v.composing, popupOpen = v.popupOpen, submitOnEnter = v.submitOnEnter),
                v.id,
            )
        }
    }

    @Test fun theImeGetsEveryKeyWhileACompositionIsOpen() {
        // The guard is unconditional on purpose: a composing Tab must not move
        // focus out from under a half-typed word, and Escape belongs to the IME.
        for (key in listOf("Enter", "Tab", "Escape", "ArrowDown", "ArrowUp", "Home", "End", "PageDown", "PageUp")) {
            assertEquals(
                FlareFormKeyboardIntent.None,
                resolveFormKeyboardIntent(key, shift = true, composing = true, popupOpen = true, submitOnEnter = true),
                key,
            )
        }
    }

    @Test fun aCappedFieldClampsInsteadOfDroppingTheUpdate() {
        // Dropping left the IME composing against a value the state no longer
        // held, and CJK input lost characters at the boundary.
        assertEquals("晚点同步", composerTextWithinLimit("晚点同步", null))
        assertEquals("晚点同步", composerTextWithinLimit("晚点同步", 8))
        assertEquals("晚点", composerTextWithinLimit("晚点同步", 2))
        assertEquals("", composerTextWithinLimit("晚点同步", 0))
    }

    @Test fun theKeyNameIsCaseInsensitive() {
        assertEquals(FlareFormKeyboardIntent.Submit, resolveFormKeyboardIntent("enter", submitOnEnter = true))
        assertEquals(FlareFormKeyboardIntent.NextOption, resolveFormKeyboardIntent("ARROWDOWN", popupOpen = true))
    }
}
